types/async-eventemitter/index.d.ts

```diff
@@ -0,0 +1,87 @@
+// Type definitions for async-eventemitter 0.2
+// Project: https://www.npmjs.com/package/async-eventemitter (Does not have to be to GitHub, but prefer linking to a source code repository rather than to a project website.)
+// Definitions by: patarapolw <https://github.com/patarapolw>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 3.5
+
+import { EventEmitter } from "events";
+
+/**
+ * The API and behavior of AsyncEventEmitter is as far as possible and meaningful identical to
+ * that of the native EventEmitter. However there are some important differences which should be noted.
+ * - Data sent to event listeners (eg emit(data)) must always be zero or one argument, and can not be a function.
+ * - Event listeners will always recieve the data object, which may or may not be undefined.
+ * - The second argument can only be a callback, and will only be supplied if the event listener has an arity of two or more (eg function(e, next){}).
+ * - Event listeners with an arity of one or zero (eg without a callback argument specified) will be treated as synchronous.
+ * - Even if all event listeners are synchronous, they will still be executed asynchronously (through setImmediate) and thus code suceeding .emit() will be executed before any event listeners.
+ * - Interupt the callback chain in async listeners by calling the callback with the error as the first parameter; in sync listeners by throwing an Error.
+ * @see https://www.npmjs.com/package/async-eventemitter#important-differences-between-asynceventemitter-the-native-eventemitter
+ */
+export type AsyncListener<T, R> = (data?: T, callback?: (result?: R) => void) => Promise<R> | void;
```

Comment:
As discussed [here](http://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html), it's not recommended to have optional parameters in callbacks.

Reply:
This will break,

```typescript
'premature-resolve': (data: number, resolve?: (result: any) => void) => Promise<any>;
...
ee.on('premature-resolve', async (data, resolve) => {
  if (resolve) resolve(true);
  'does not reach here';
});
```

That it should not require anything in resolve (default to `undefined`).

Reply:
Ah, I don't think I was clear.
If I make the following change, then everything works as expected, including the lines you've specified in your comment:
```ts
export type AsyncListener<T, R> = (data: T, callback: (result?: R) => void) => Promise<R> | void;
```

In this case, the callback I'm concerned about is the whole `AsyncListener`. The `?` on `result` is correct.